src/main/scala/io/picnicml/doddlemodel/preprocessing/RangeScaler.scala

```diff
@@ -0,0 +1,58 @@
+package io.picnicml.doddlemodel.preprocessing
+
+import breeze.linalg.{*, Axis, max, min}
+import cats.syntax.option._
+import io.picnicml.doddlemodel.data.Feature.FeatureIndex
+import io.picnicml.doddlemodel.data.{Features, RealVector}
+import io.picnicml.doddlemodel.syntax.OptionSyntax._
+import io.picnicml.doddlemodel.typeclasses.Transformer
+
+case class RangeScaler private (private val scale: Option[RealVector],
+                                private val minAdjustment: Option[RealVector],
+                                private val range: (Double, Double),
+                                private val featureIndex: FeatureIndex)
+
+object RangeScaler {
+
+  def apply(range: (Double, Double), featureIndex: FeatureIndex): RangeScaler = {
+    val (lowerBound, upperBound) = range
+    require(upperBound > lowerBound, "Upper bound of range must be greater than lower bound")
+    RangeScaler(none, none, range, featureIndex)
+  }
+
+  implicit lazy val ev: Transformer[RangeScaler] = new Transformer[RangeScaler] {
+
+    override def isFitted(model: RangeScaler): Boolean =
+      model.scale.isDefined && model.minAdjustment.isDefined
+
+    override def fit(model: RangeScaler, x: Features): RangeScaler = {
+      val (lowerBound, upperBound) = model.range
+      val numericColIndices = model.featureIndex.numerical.columnIndices
+      val numericColsOnly = x(::, numericColIndices).toDenseMatrix
```

Comment:
I think we can simply remove `.toDenseMatrix` here and then write `(max(x(::, numericColIndices), Axis._0).inner, min(x(::, numericColIndices), Axis._0).inner)`.

Reply:
Good catch. One small problem remains though - the line `val dataRange = colMax - colMin` somehow creates a `Vector[Double]` out of two `DenseVector[Double]`s, making the following line fail (because `:==` is not possible between breeze's `Vector` and `Double`):
`dataRange(dataRange :== 0.0) := 1.0`.

I see two ways to fix this, please let me know which one sounds better/ if there's a better way (seems like there should be):
1.) Type cast, i.e. `val dataRange = (colMax - colMin).asInstanceOf[DenseVector[Double]]`  
- does not look very nice, pretty sure explicit type casts are to be avoided if possible
- not sure how this is with copying (how much copying does it do, I don't see how it could be in-place)

2.) `toDenseVector`  
- same problems as `toDenseMatrix`, but vectors are 1D and therefore have a smaller memory footprint, so not as much data would be copied.

Reply:
Hmm `val dataRange = colMax - colMin` actually results in `dataRange: DenseVector[Double]` as far as I can see (tests are also green). Try running the code and see if you actually observe `dataRange: Vector[Double]` (maybe it's reported by your IDE wrongly).

Reply:
It seems you were right, it works. I was not aware I could be tricked by the REPL as well as the IDE :sweat_smile:

Reply:
Weird indeed. Which IDE are you using?

Reply:
A quite stale version of IntelliJ Ultimate edition (2018.2.4). It seems like it's time to finally update it :sweat_smile:

Reply:
Ok, this is the only thing that still doesn't compile for me 😅. This compiles, however:
```scala
val colMax = max(x(::, numericColIndices), Axis._0).t.toDenseVector
val colMin = min(x(::, numericColIndices), Axis._0).t.toDenseVector
val dataRange = colMax - colMin
```
it's still better than the previous `.toDenseMatrix` as it only materializes two vectors of length `x(::, numericColIndices).cols`. I belive the max and min actually returns a `Vector[Double]` because we are passing in a `SliceMatrix`. Anyway, the above should work as intended.

Reply:
Fixed. Weird that the tests passed anyway, are you using a version of breeze that is different from the one in project/CI? Anyways, good that you caught the problem before merging :smiley:

Reply:
That shouldn't be the case as I checked out your branch. I'm not sure why it didn't fail, I will have to investigate 🙂.